pkg/suggestion/v1alpha3/goptuna/service.go

```diff
@@ -0,0 +1,129 @@
+)
+
+const (
+	AlgorithmCMAES  = "cmaes"
+	AlgorithmTPE    = "tpe"
+	AlgorithmRandom = "random"
+)
+
+func NewSuggestionService() *SuggestionService {
+	return &SuggestionService{}
+}
+
+type SuggestionService struct{}
+
+func (s *SuggestionService) GetSuggestions(
+	ctx context.Context,
+	req *api_v1_alpha3.GetSuggestionsRequest,
+) (*api_v1_alpha3.GetSuggestionsReply, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "request is empty")
+	}
+
+	direction, err := toGoptunaDirection(req.GetExperiment().GetSpec().GetObjective().GetType())
+	if err != nil {
+		klog.Errorf("Failed to convert to Goptuna direction: %s", err)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	independentSampler, relativeSampler, err := toGoptunaSampler(req.GetExperiment().GetSpec().GetAlgorithm())
+	if err != nil {
+		klog.Errorf("Failed to create Goptuna sampler: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	searchSpace, err := toGoptunaSearchSpace(req.GetExperiment().GetSpec().GetParameterSpecs().GetParameters())
+	if err != nil {
+		klog.Errorf("Failed to convert to Goptuna search space: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	klog.Infof("Goptuna search space: %#v", searchSpace)
+
+	studyOpts := make([]goptuna.StudyOption, 0, 3)
+	studyOpts = append(studyOpts, goptuna.StudyOptionSetDirection(direction))
+	if independentSampler != nil {
+		studyOpts = append(studyOpts, goptuna.StudyOptionSampler(independentSampler))
+	}
+	if relativeSampler != nil {
+		studyOpts = append(studyOpts, goptuna.StudyOptionRelativeSampler(relativeSampler))
+	}
+
+	study, err := goptuna.CreateStudy("katib", studyOpts...)
+	if err != nil {
+		klog.Errorf("Failed to create Goptuna study: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	trials, err := toGoptunaTrials(req.GetTrials(), study, searchSpace)
+	if err != nil {
+		klog.Errorf("Failed to convert to Goptuna trials: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	for _, t := range trials {
+		_, err = study.Storage.CloneTrial(study.ID, t)
+		if err != nil {
+			klog.Errorf("Failed to register trials: %s", err)
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	requestNumber := int(req.GetRequestNumber())
+	parameterAssignments := make([]*api_v1_alpha3.GetSuggestionsReply_ParameterAssignments, requestNumber)
+	for i := 0; i < requestNumber; i++ {
+		assignments, err := sampleNextParam(study, searchSpace)
+		if err != nil {
+			klog.Errorf("Failed to sample next param: %s", err)
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		parameterAssignments[i] = &api_v1_alpha3.GetSuggestionsReply_ParameterAssignments{
+			Assignments: assignments,
+		}
+	}
+
+	klog.Infof("Success to sample %d parameters", requestNumber)
+	return &api_v1_alpha3.GetSuggestionsReply{
+		ParameterAssignments: parameterAssignments,
+		Algorithm: &api_v1_alpha3.AlgorithmSpec{
```

Comment:
Do you need to send `Algorithm` in `GetSuggestionsReply`, is it `Golang` requirements?

Reply:
Thank you for you comment. It doesn't need to send `Algorithm`.
Fixed at 4522744b.